simln-lib/src/sim_node.rs

```diff
@@ -1896,4 +2198,168 @@ mod tests {
         test_kit.graph.tasks.wait().await;
     }
+
+    /// Tests intercepted htlc failures.
+    #[tokio::test]
+    async fn test_intercepted_htlc_failure() {
+        // Test with 2 interceptors where one of them returns a signal to fail the htlc.
+        let mut mock_interceptor_1 = MockTestInterceptor::new();
+        mock_interceptor_1
+            .expect_intercept_htlc()
+            .returning(|_| Ok(Ok(CustomRecords::default())));
+        mock_interceptor_1
+            .expect_notify_resolution()
+            .returning(|_| Ok(()));
+
+        let mut mock_interceptor_2 = MockTestInterceptor::new();
+        mock_interceptor_2.expect_intercept_htlc().returning(|_| {
+            Ok(Err(ForwardingError::InterceptorError(
+                "failing from mock interceptor".into(),
+            )))
+        });
+        mock_interceptor_2
+            .expect_notify_resolution()
+            .returning(|_| Ok(()));
+
+        let chan_capacity = 500_000_000;
+        let mock_1 = Arc::new(mock_interceptor_1);
+        let mock_2 = Arc::new(mock_interceptor_2);
+        let mut test_kit = DispatchPaymentTestKit::new(chan_capacity, vec![mock_1, mock_2]).await;
+        let expected_balances = vec![(chan_capacity, 0), (chan_capacity, 0), (chan_capacity, 0)];
+
+        // Send payment where there is enough liquidity but one of the interceptors fails the htlc.
+        let amt = 150_000;
+        let (_, result) = test_kit
+            .send_test_payment(test_kit.nodes[0], test_kit.nodes[3], amt)
+            .await;
+
+        assert_eq!(test_kit.channel_balances().await, expected_balances);
+        assert!(matches!(
+            result.unwrap().payment_outcome,
+            PaymentOutcome::Unknown
+        ));
+
+        let mut mock_interceptor_1 = MockTestInterceptor::new();
+        mock_interceptor_1
+            .expect_intercept_htlc()
+            .returning(|_| Ok(Ok(CustomRecords::from([(1000, vec![1])]))));
+        mock_interceptor_1
+            .expect_notify_resolution()
+            .returning(|_| Ok(()));
+
+        let mut mock_interceptor_2 = MockTestInterceptor::new();
+        mock_interceptor_2
+            .expect_intercept_htlc()
+            .returning(|_| Ok(Ok(CustomRecords::from([(1000, vec![1])]))));
+        mock_interceptor_2
+            .expect_notify_resolution()
+            .returning(|_| Ok(()));
+
+        let mock_1 = Arc::new(mock_interceptor_1);
+        let mock_2 = Arc::new(mock_interceptor_2);
+        let mut test_kit = DispatchPaymentTestKit::new(chan_capacity, vec![mock_1, mock_2]).await;
+
+        // Test intercepted htlc with conflicting records. Conflicting records should fail the
+        // htlc.
+        let (_, result) = test_kit
+            .send_test_payment(test_kit.nodes[0], test_kit.nodes[3], amt)
+            .await;
+
+        assert!(matches!(
+            result.unwrap().payment_outcome,
+            PaymentOutcome::Unknown
+        ));
+
```

Comment:
Right now I can make this test fail by adding a `listener` to the test kit and checking that we've triggered top level shutdown of the simulation (because a failed interceptor shuts everything down).

```
		select!{
			_ = test_kit.listener => {assert!(false, "simulation shut down by interceptor")},
			_ = async {} => {},
		}
```

Would be good to add this in above here to assert that interceptor failures aren't killing the whole simulation.